### PR TITLE
Use built-in CMake support for iOS

### PR DESCRIPTION
**Pull request by JonnyPtn**
Now we're using more modern cmake, iOS support is first-class so we don't need to ship our own toolchain.

Practically there should be no difference except for generally being easier to work with iOS (e.g don't need to reconfigure to switch between simulator/device)

### CODE REVIEW

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: cmake/Config.cmake [hunk 1]
```diff
@@ -39,21 +39,22 @@ elseif(CMAKE_SYSTEM_NAME MATCHES "^NetBSD$")
...
     # don't use the OpenGL ES implementation on NetBSD
     set(OPENGL_ES 0)
+elseif(${CMAKE_SYSTEM_NAME} STREQUAL "iOS")
+    set(SFML_OS_IOS 1)
+    # As we want to find packages in our extlibs folder too
+    # we need to tell CMake to search there as well
+    set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY NEVER)
+    set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE NEVER)
+    set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE NEVER)
+    set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
```
Comment (eXpl0it3r) [inline cmake/Config.cmake:49]: The comment says "our extlibs folder too", but then the option is set to `NEVER`, doesn't that mean, we're exclusively searching the `extlibs` folder?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -38,7 +38,7 @@ jobs:
         - platform: { name: MacOS, os: macos-11 }
           config: { name: Frameworks, flags: -DSFML_BUILD_FRAMEWORKS=TRUE }
         - platform: { name: MacOS, os: macos-11 }
-          config: { name: iOS, flags: -DCMAKE_TOOLCHAIN_FILE=$GITHUB_WORKSPACE/cmake/toolchains/iOS.toolchain.cmake -DIOS_PLATFORM=SIMULATOR }
+          config: { name: iOS, flags:  -DCMAKE_SYSTEM_NAME=iOS }
```
Comment (ChrisThrasher) [inline .github/workflows/ci.yml:41]:
```suggestion
          config: { name: iOS, flags: -DCMAKE_SYSTEM_NAME=iOS }
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -12,15 +12,7 @@ endmacro()
...
-
-if(NOT CMAKE_OSX_SYSROOT)
-    # query the path to the default SDK, will fail on non-macOS, but it's okay.
-    execute_process(COMMAND xcodebuild -sdk macosx -version Path
-                    COMMAND head -n 1
-                    COMMAND tr -d '\n'
-                    OUTPUT_VARIABLE CMAKE_OSX_SYSROOT
-                    ERROR_QUIET)
-endif()
+sfml_set_option(CMAKE_OSX_DEPLOYMENT_TARGET "13.0" STRING "The minimal iOS version that will be able to run the built binaries. Cannot be lower than 13.0")
```
Comment (ChrisThrasher) [inline CMakeLists.txt:15]: Can we move this somewhere after the `project` call? I'd like to minimum how much code we add before `project()`.

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
I'm not familiar with CMake for iOS but if this works then it looks like a great simplification.

File: CMakeLists.txt (hunk 3 above)
Comment (JonnyPtn) [reply CMakeLists.txt:15]: Nope it has to be before the first call to project https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_DEPLOYMENT_TARGET.html

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: cmake/Config.cmake (hunk 1 above)
Comment (JonnyPtn) [reply cmake/Config.cmake:49]: Yeah good spot - originally set to BOTH but then changed to NEVER, will update comment

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML-Buildbot/blob/0ab146c329327a85a191f716e8de41c8070f7325/buildfactories.py#L67-L69

Looks like we need to update the Buildbot too

**Comment by JonnyPtn** [comment]
Think I've done all I can/should on this for now - Ended up going down a bit of a rabbit hole with our Xcode/frameworks stuff which is a bit FUBAR right now, so this just adds the tests for iOS static libraries right now and we can address the framework issues separately as there is some overlap with macOS

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
Being able to get rid of that iOS toolchain and test the install interface on iOS is promising. I've got a few more clarifying questions since idk much about iOS development.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -112,7 +118,7 @@ jobs:
       shell: bash
       run: |
         cmake -S $GITHUB_WORKSPACE/test/install -B $GITHUB_WORKSPACE/test/install/build -DSFML_ROOT=$GITHUB_WORKSPACE/install -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=${{ matrix.type.name == 'Debug' && 'Debug' || 'Release' }} ${{matrix.platform.flags}} ${{matrix.config.flags}} ${{matrix.type.flags}}
-        cmake --build $GITHUB_WORKSPACE/test/install/build --config ${{ matrix.type.name == 'Debug' && 'Debug' || 'Release' }}
+        cmake --build $GITHUB_WORKSPACE/test/install/build --config ${{ matrix.type.name == 'Debug' && 'Debug' || 'Release' }} ${{matrix.platform.suffix}}
```
[the remark this answers is not in the document]
Comment (JonnyPtn) [reply .github/workflows/ci.yml:121]: Ah, they should be running - probably got lost in the crossfire when I was trying to get the framework tests running, will amend

File: examples/assets/info.plist [hunk 5]
```diff
@@ -23,11 +23,11 @@
 	<key>CFBundlePackageType</key>
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string></string>
+	<string>2.5.1</string>
```
[the remark this answers is not in the document]
Comment (JonnyPtn) [reply examples/assets/info.plist:26]: I don't recall, probably just left over from when I first looked at this, will amend

File: src/SFML/Window/CMakeLists.txt [hunk 6]
```diff
@@ -342,7 +340,7 @@ elseif(SFML_OS_FREEBSD)
 elseif(SFML_OS_MACOSX)
     target_link_libraries(sfml-window PRIVATE "-framework Foundation" "-framework AppKit" "-framework IOKit" "-framework Carbon")
 elseif(SFML_OS_IOS)
-    target_link_libraries(sfml-window PRIVATE "-framework Foundation" "-framework UIKit" "-framework CoreGraphics" "-framework QuartzCore" "-framework CoreMotion")
+    target_link_libraries(sfml-window PUBLIC "-framework Foundation" "-framework UIKit" "-framework CoreGraphics" "-framework QuartzCore" "-framework CoreMotion")
```
[the remark this answers is not in the document]
Comment (JonnyPtn) [reply src/SFML/Window/CMakeLists.txt:343]: The frameworks need to be linked by the dependent build. I believe this is propagated automatically if we use find_package along with proper targets but as we just use dumb compiler flags we have to do this (or require that anybody using it links the frameworks themselves)

File: test/install/CMakeLists.txt [hunk 7]
```diff
@@ -8,8 +8,16 @@ if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
     if(NOT BUILD_SHARED_LIBS)
         set(SFML_STATIC_LIBRARIES TRUE)
     endif()
+
+    # cmake by default will only search inside the iOS SDK for packages/libraries, so we need to tell it to look elsewhere
+    if (IOS)
+        set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE BOTH)
+        set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY BOTH)
+    endif()
```
[the remark this answers is not in the document]
Comment (JonnyPtn) [reply test/install/CMakeLists.txt:16]: Yeah as above - I've fixed the issues with the install tests (except for frameworks which is a separate can of worms) so need to enable the tests

File: test/install/CMakeLists.txt [hunk 8]
```diff
@@ -8,8 +8,16 @@ if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
...
+
     find_package(SFML 3.0.0 EXACT CONFIG REQUIRED COMPONENTS Graphics Network Audio)
 endif()
 
 add_executable(test-sfml-install Install.cpp)
 target_link_libraries(test-sfml-install PRIVATE SFML::Graphics SFML::Network SFML::Audio)
+message(STATUS ${CMAKE_OSX_SYSROOT})
```
Comment (ChrisThrasher) [inline test/install/CMakeLists.txt:23]: Why this message?

File: test/install/CMakeLists.txt [hunk 9]
```diff
@@ -8,8 +8,16 @@ if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
     if(NOT BUILD_SHARED_LIBS)
         set(SFML_STATIC_LIBRARIES TRUE)
     endif()
+
+    # cmake by default will only search inside the iOS SDK for packages/libraries, so we need to tell it to look elsewhere
+    if (IOS)
```
Comment (ChrisThrasher) [inline test/install/CMakeLists.txt:13]:
```suggestion
    if(IOS)
```

**Review by ChrisThrasher — APPROVED** [review]
Other than two minor comments, I like this! I can tell this is going to unblock some things I've previously tried to modernize how we handle dependencies. I'm eager to try those again once this is merged.

File: test/install/CMakeLists.txt (hunk 8 above)
Comment (JonnyPtn) [reply test/install/CMakeLists.txt:23]: It was mostly useful for diagnosing SDK compatibility issues so you can see which one is being used - happy to take any direction on whether it should be removed or formatted differently

**Comment by eXpl0it3r** [comment]
Looks like, we'll need to adapt BuildBot afterwards

**Comment by eXpl0it3r** [comment]
Created a change for Build Bot, but I'm unsure if we can support this both for master and 2.6.x based branches: https://github.com/SFML/SFML-Buildbot/pull/39

**Comment by binary1248** [comment]
> Created a change for Build Bot, but I'm unsure if we can support this both for master and 2.6.x based branches: [SFML/SFML-Buildbot#39](https://github.com/SFML/SFML-Buildbot/pull/39)

This is possible... and has been implemented.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]

**Comment by eXpl0it3r** [comment]
Thanks for cleaning this up! 🧹🪣

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/blob/3f6403e279d3f0965938250f263cdd9fdfef5f4a/cmake/Macros.cmake#L35-L36

Should these lines be removed? They expand into nothing and due to how `set_property` works that results in technically valid (albeit not useful) CMake code.

**Comment by JonnyPtn** [comment]
From memory, they existed so that people could specify the properties themselves on the command line when building SFML - Can still be done with the global CMAKE_XCODE_ATTRIBUTE_... variables if these were removed, but then it wouldn't be possible to have different settings for different targets.

Benefit and cost are both very small so not sure it really matters much

**Pull request merged**